design/design-generic-references.md

```diff
@@ -0,0 +1,222 @@
+# Generic Resource References
+
+* Owner: Muvaffak Onuş (@muvaf)
+* Reviewers: Crossplane Maintainers
+* Status: Accepted
+
+## Background
+
+In many cases, Crossplane managed resources require information that is not
+available when the user creates the resource. One example is that if you'd like to create
+a `Subnet`, you need to supply a VPC ID but if you create the `VPC` at the same
+time then the ID will be available only after the creation, so you need to extract
+the ID and use it in `Subnet`. Today, it's possible to let the provider do this
+operation using cross-resource references. However, this requires knowing the kind of
+the referenced resource and how to extract the value during development time of
+the provider. There are several cases where this is not possible:
+
+* The referencer field may not always target a single kind. For example, a URL
+  field in Route53 could get populated by a URL of an S3 bucket or access endpoint
+  of an RDS Instance.
+* Instead of hard-coding region for all managed resources, you may want to refer
+  to a `ConfigMap` or a `Secret` and let the managed resource take that
+  information from there.
+* In Composition, you can use values of the fields only between the children of
+  the same composite instance, but there are cases where you'd like to use a
+  separate resource that is not bound to any composite instance to store the
+  information and serve all composites.
+
```

Comment:
Wanted to offer another use-case that we've run into:

* Enabling a Workload Identity on a GKE Cluster requires supplying a value which contains the Project ID which hosts the cluster.  Because compositions are intended to be re-used, this requires passing in the Project ID at render time. It _is_ possible to add this as a parameter to the Composition, but this is information that a Claim Author is not expected to manage or know. A Platform Operator would want to supply this data " behind-the-scenes." In this case, allowing Composition Authors to source a `ConfigMap` or `Secret` outside the Composition would allow Platform Operators to add manage these values without Claim Authors needing to take on extra context.